moveit_kinematics/cached_ik_kinematics_plugin/src/ik_cache.cpp

```diff
@@ -0,0 +1,341 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2017, Rice University
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the Rice University nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Mark Moll */
+
+#include <boost/filesystem/fstream.hpp>
+#include <chrono>
+#include <cstdlib>
+#include <numeric>
+
+#include <moveit/cached_ik_kinematics_plugin/cached_ik_kinematics_plugin.h>
+
+namespace cached_ik_kinematics_plugin
+{
+IKCache::IKCache()
+{
+  // set distance function for nearest-neighbor queries
+  ikNN_.setDistanceFunction([this](const IKEntry* entry1, const IKEntry* entry2) {
+    double dist = 0.;
+    for (unsigned int i = 0; i < entry1->first.size(); ++i)
+      dist += entry1->first[i].distance(entry2->first[i]);
+    return dist;
+  });
+}
+
+IKCache::~IKCache()
+{
+  if (!ikCache_.empty())
+    saveCache();
+}
+
+void IKCache::initializeCache(const std::string& robot_description, const std::string& group_name,
+                              const std::string& cache_name, const unsigned int num_joints)
+{
+  // read ROS parameters
+  ros::NodeHandle node_handle("~");
+  int maxCacheSize;  // node_handle.param(...) doesn't handle unsigned ints
```

Comment:
inconsistent use of camel case vs underscore variables. underscore is our standard.

Reply:
Will fix.